### PR TITLE
Custom build steps to fetch CFamily files and package in VSIX

### LINKED ISSUE #929: Modify build to acquire and package the correct C++ analyzer exe
#### Notes
We need to embed the following:
* the subprocess exe
* the json files specify which rules to run and the parameters for those rules
* ?a set of `.h` files? These files are only required for ObjectiveC analysis, so we should be able to avoid using them
* the following `.json` files: Sonar_way_profile.json, RulesList.json, and all of the XXX_params.json files

All of the above are embedded in the CFamily plugin jar, which is publicly available at https://binaries.sonarsource.com/CommercialDistribution/sonar-cfamily-plugin/.

The simplest solution will be to download the whole jar, unzip it, and extract the bits we need.
This would also mean that anyone forking this repo will still be able to build SLVS, even though the CFamily jar is commercially licensed. It also means that we don't need to worry about having to provide credentials in our internal builds to pull the bits from non-public locations.

To enable testing of intermediate builds, it would be useful if the download location and CFamily plugin version could be parameterised to allow developers to set them locally.

#### TODO
- [ ] modify the build to download the jar from https://binaries.sonarsource.com/CommercialDistribution/sonar-cfamily-plugin/
- [ ] package the jar and json files into the VSIX. Check the new VSIX size
- [ ] test on Azure Pipeline build
- [ ] check the increase in build time for PR/CI builds. If the build time increases too much then it might be necessary to skip the CFamily packaging step in PR builds.

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixes #929

### CODE REVIEW

File: build/DownloadCFamilyPlugin/DownloadAndExtract.cs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    /// <remarks>
+    /// Assumptions:
+    /// * we are downloading the CFamily plugin jar e.g. https://binaries.sonarsource.com/CommercialDistribution/sonar-cfamily-plugin/sonar-cfamily-plugin-6.3.0.11371.jar
+    /// * the jar contains a .tarxz archive containing the Windows version of SubProcess.exe
+    /// * the jar contains various .json files, some with fixed names, some that we'll match using wildcards
+    /// 
+    /// We want the task to fail if it cannot locate all of the expected files (otherwise we'll build an invalid VSIX).
+    /// 
+    /// Downloading and extracting the files can be slow so we want to skip those steps if possible.
+    /// </remarks>
```
Comment (duncanp-sonar) [inline build/DownloadCFamilyPlugin/DownloadAndExtract.cs]: There's quite a bit of processing required to handle fetching and embedding the files to I created a custom MSBuild task. This is in a new project under the existing `build` folder.

File: build/DownloadCFamilyPlugin/DownloadAndExtract.cs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+        private readonly string[] KnownJsonFiles = new string[]
+        {
+                "Sonar_way_profile.json",
+                "RulesList.json"
+        };
+
+        private readonly string[] AdditionalFilesToReturnPattern = new string[]
+        {
+            "S*_params.json"
```
Comment (henryju) [inline build/DownloadCFamilyPlugin/DownloadAndExtract.cs]: Not all rules starts with 'S': https://github.com/SonarSource/sonar-cpp/tree/master/cpp-checks/src/main/resources/org/sonar/l10n/cpp/rules/params

File: build/DownloadCFamilyPlugin/DownloadCFamilyPlugin.csproj [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net46</TargetFramework>
+    
+    <!-- Not a product project so we don't want it to be taken into account for code coverage -->
```
Comment (henryju) [inline build/DownloadCFamilyPlugin/DownloadCFamilyPlugin.csproj]: In the long run, we may even want to completely exclude it from SQ analysis.

**Comment by henryju** [comment]
Will this way of packaging the files allow to use them in unit tests? I tried but seems the files are not in the bin folder.

File: build/DownloadCFamilyPlugin/DownloadCFamilyPlugin.csproj (hunk 3 above)
Comment (duncanp-sonar) [reply build/DownloadCFamilyPlugin/DownloadCFamilyPlugin.csproj]: Good point point. I've changed the property to `<SonarQubeExclude>true</SonarQubeExclude> so the entire project will be completely ignored.

**Comment by duncanp-sonar** [comment]
Superceded.

**Pull request closed without merging**